### PR TITLE
feat: Bearer token auth

### PR DESCRIPTION
**Pull request by sichanyoo**
Companion PR: https://github.com/smithy-lang/smithy-swift/pull/786

## Issue \#

https://github.com/awslabs/aws-sdk-swift/issues/1083

## Description of changes

Refer to comments on github diff

## Manual testing steps
To test this locally, first set up your IAM Identity Center ([doc link](https://docs.aws.amazon.com/singlesignon/latest/userguide/get-set-up-for-idc.html)) & AWS CodeCatalyst ([doc link](https://docs.aws.amazon.com/codecatalyst/latest/userguide/setting-up-topnode.html)).
Then, configure your config file to contain the following:
```
[default]
region=us-west-2
output=json
sso_session = codecatalyst

[profile codecatalyst]
region = us-west-2
sso_session = codecatalyst

[sso-session codecatalyst]
sso_region = us-east-1
sso_start_url = https://view.awsapps.com/start
sso_registration_scopes = codecatalyst:read_write
```
Then, run 
```
aws sso login --profile=codecatalyst
```
on the terminal. This will open a browser tab for you to authorize your AWS account to be accessed by CodeCatalyst.
Successful previous step creates a token file at path `~/.aws/sso/cache/<sha-1-hash-of-sso-sesion-name>.json`.
In CLITool, use `CodeCatalyst::VerifySession` API by simply doing the following:
```
func run() async throws {
        let client = try CodeCatalystClient(region: "us-west-2")
        let id = try await client.verifySession(input: VerifySessionInput()).identity
        print(id ?? "#### NO_ID_RETRIEVED")
}
```
Observe that the server accepts bearer auth and returns the ID successfully.

## Conventional Commits

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: Sources/Core/AWSSDKIdentity/Sources/AWSSDKIdentity/BearerTokenIdentityResolvers/DefaultBearerTokenIdentityResolverChain.swift [hunk 1]
```diff
@@ -0,0 +1,35 @@
+//
```
Comment (sichanyoo) [inline Sources/Core/AWSSDKIdentity/Sources/AWSSDKIdentity/BearerTokenIdentityResolvers/DefaultBearerTokenIdentityResolverChain.swift:1]: This is much like the default AWS credential identity resolver chain. Unlike it tho, this only has one resolver atm.

File: Sources/Core/AWSSDKIdentity/Sources/AWSSDKIdentity/BearerTokenIdentityResolvers/SSOBearerTokenIdentityResolver.swift [hunk 2]
```diff
@@ -0,0 +1,94 @@
+//
```
Comment (sichanyoo) [inline Sources/Core/AWSSDKIdentity/Sources/AWSSDKIdentity/BearerTokenIdentityResolvers/SSOBearerTokenIdentityResolver.swift:1]: This token identity resolver fetches the access token from cached SSO token file.

File: codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/AWSAuthUtils.kt [hunk 3]
```diff
@@ -18,8 +18,11 @@ import software.amazon.smithy.swift.codegen.SwiftWriter
 import software.amazon.smithy.swift.codegen.integration.ProtocolGenerator
```
Comment (sichanyoo) [inline codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/AWSAuthUtils.kt:18]: Refactored to remove logic for getting auth schemes supported by the model and move it into `smithy-swift`. Refer to comment: https://github.com/smithy-lang/smithy-swift/pull/786/files#r1681867714

File: codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/AWSAuthUtils.kt [hunk 4]
```diff
@@ -57,24 +60,22 @@ open class SigV4Utils() {
...
-            }
-            return "[${authSchemeList.joinToString(", ")}]"
+        if (effectiveAuthSchemes.contains(SigV4Trait.ID)) {
+            updatedAuthSchemeList += writer.format("\$N()", AWSSDKHTTPAuthTypes.SigV4AuthScheme)
+        }
+        if (effectiveAuthSchemes.contains(SigV4ATrait.ID) || servicesUsingSigV4A.contains(sdkId)) {
+            updatedAuthSchemeList += writer.format("\$N()", AWSSDKHTTPAuthTypes.SigV4AAuthScheme)
         }
+
+        return updatedAuthSchemeList
```
Comment (sichanyoo) [inline codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/AWSAuthUtils.kt:79]: Handles detecting and adding the SigV4 / SigV4A auth schemes to list of auth schemes supported by the service.

File: codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/AWSHttpProtocolServiceClient.kt [hunk 5]
```diff
@@ -45,7 +47,15 @@ class AWSHttpProtocolServiceClient(
...
-                    ConfigProperty("authSchemes", SmithyHTTPAuthAPITypes.AuthSchemes.toOptional(), authSchemesDefaultProvider)
+                    ConfigProperty("authSchemes", SmithyHTTPAuthAPITypes.AuthSchemes.toOptional(), AWSAuthUtils(ctx).authSchemesDefaultProvider)
+                }
+                "bearerTokenIdentityResolver" -> {
+                    ConfigProperty(
+                        "bearerTokenIdentityResolver",
+                        SmithyIdentityTypes.BearerTokenIdentityResolver.toGeneric(),
+                        { it.format("\$N()", AWSSDKIdentityTypes.DefaultBearerTokenIdentityResolverChain) },
+                        true
+                    )
```
Comment (sichanyoo) [inline codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/AWSHttpProtocolServiceClient.kt:58]: The default value for `bearerTokenIdentityResolver` field of the client config was `StaticBearerTokenIdentityResolver` in `smithy-swift`, but over here in `aws-sdk-swift`, it has to be the default chain instead. Hence this override.

File: codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/AWSHttpProtocolServiceClient.kt [hunk 6]
```diff
@@ -130,12 +140,6 @@ class AWSHttpProtocolServiceClient(
         }
     }
 
-    private val authSchemesDefaultProvider = DefaultProvider(
-        { getModeledAuthSchemesSupportedBySDK(ctx, it) },
-        isThrowable = false,
-        isAsync = false
-    )
-
```
Comment (sichanyoo) [inline codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/AWSHttpProtocolServiceClient.kt:138]: Moved to `AuthUtils` in `smithy-swift`.

File: codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/plugins/AuthSchemePlugin.kt [hunk 7]
```diff
@@ -28,16 +28,20 @@ class AuthSchemePlugin(private val serviceConfig: ServiceConfig) : Plugin {
             writer.write("private var authSchemes: \$N", SmithyHTTPAuthAPITypes.AuthSchemes.toOptional())
```
Comment (sichanyoo) [inline codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/plugins/AuthSchemePlugin.kt:28]: Changes in this file are just adding `bearerTokenIdentityResolver` field as one of the new fields configurable by the auth scheme plugin.

File: codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/plugins/DefaultAWSAuthSchemePlugin.kt [hunk 8]
```diff
@@ -1,7 +1,8 @@
 package software.amazon.smithy.aws.swift.codegen.plugins
```
Comment (sichanyoo) [inline codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/plugins/DefaultAWSAuthSchemePlugin.kt:2]: Changes in this file are just adding `bearerTokenIdentityResolver` field as one of the new fields configurable by the auth scheme plugin.

File: codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/swiftmodules/AWSSDKIdentityTypes.kt [hunk 9]
```diff
@@ -18,6 +18,7 @@ object AWSSDKIdentityTypes {
     val AWSEndpointsRequestContext = runtimeSymbol("AWSEndpointsRequestContext")
```
Comment (sichanyoo) [inline codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/swiftmodules/AWSSDKIdentityTypes.kt:18]: The new runtime type, but in codegen side.

**Review by sichanyoo — COMMENTED** [review]
Comments to help reviewers.

**Review by dayaffe — APPROVED** [review]

**Pull request merged**